### PR TITLE
Compute disk usage stats recursively for all folders

### PR DESCRIPTION
**Pull request by lgruen**
(Except for `.ht` / `.mt`.)

I'd like to build a simple treemap visualization based on this.

### CODE REVIEW

File: disk_usage/disk_usage.py [hunk 1]
```diff
@@ -27,40 +28,33 @@
...
 
-    aggregate_stats = defaultdict(AggregateStats)
+    aggregate_stats = defaultdict(lambda: defaultdict(int))
     for bucket_suffix in BUCKET_SUFFIXES:
+        if access_level == 'test' and not bucket_suffix.startswith('test'):
+            continue  # Skip main buckets when testing.
+
         bucket_name = f'cpg-{dataset}-{bucket_suffix}'
         logging.info(f'Listing blobs in {bucket_name}...')
         blobs = storage_client.list_blobs(bucket_name)
```
Comment (illusional) [inline disk_usage/disk_usage.py:60]: Does this not already list every single file in the bucket? I thought you had to be explicit about filtering to just the bucket root or something.

File: disk_usage/disk_usage.py [hunk 2]
```diff
@@ -27,40 +28,33 @@
 
 
 def aggregate_level(name: str) -> str:
-    """Returns a prefix for the given blob name at the aggregation level."""
+    """Returns a prefix for the given blob name at the folder or Hail table level."""
```
Comment (illusional) [inline disk_usage/disk_usage.py:31]: Should throw a couple docstring tests in here :)

File: disk_usage/disk_usage.py [hunk 3]
```diff
@@ -69,23 +63,28 @@ def main():
...
-            stats.num_blobs += 1
+            folder = aggregate_level(blob.name)
+            last_index = 0
+            while True:
+                index = folder.find('/', last_index)
+                substr = folder[:index] if index != -1 else folder
+                path = f'gs://{bucket_name}/{substr}'
+                stats = aggregate_stats[path]
+                stats['size'] += blob.size
+                stats['num_blobs'] += 1
```
Comment (illusional) [inline disk_usage/disk_usage.py:74]: Stylistically I prefer setting on the instance, but interested for this many millions if it makes a substantial difference.

```suggestion
                aggregate_stats[path]['size'] += blob.size
                aggregate_stats[path]['num_blobs'] += 1
```

File: disk_usage/disk_usage.py [hunk 4]
```diff
@@ -27,40 +28,33 @@
...
 
 def aggregate_level(name: str) -> str:
-    """Returns a prefix for the given blob name at the aggregation level."""
+    """Returns a prefix for the given blob name at the folder or Hail table level."""
     ht_index = name.find('.ht/')
     if ht_index != -1:
         return name[: ht_index + 3]
     mt_index = name.find('.mt/')
     if mt_index != -1:
         return name[: mt_index + 3]
```
Comment (illusional) [inline disk_usage/disk_usage.py:37]: You could probably use the [walrus syntax](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions) here:

```suggestion
    if (ht_index := name.find('.ht/')) != -1:
        return name[: ht_index + 3]
    if (mt_index := name.find('.mt/')) != -1:
        return name[: mt_index + 3]
```

**Review by illusional — APPROVED** [review]

File: disk_usage/disk_usage.py (hunk 3 above)
Comment (lgruen) [reply disk_usage/disk_usage.py:74]: Probably doesn't make a huge difference overall.

File: disk_usage/disk_usage.py (hunk 2 above)
Comment (lgruen) [reply disk_usage/disk_usage.py:31]: Thanks, done!

File: disk_usage/disk_usage.py (hunk 4 above)
Comment (lgruen) [reply disk_usage/disk_usage.py:37]: Nice, thanks!

**Pull request merged**